Search autocompletion.

aartaka opened the pull request:
Behold! Search suggestions in Nyxt! (see an attached screenshot)

![autocompletion](https://user-images.githubusercontent.com/57838654/117065330-cfc15c00-ad40-11eb-8d43-d32c56ae40e3.png)

This adds search autocompletion requests to the default `search-engine`s, based on the new extensible mechanism:
- `completion-function` slot of `search-engine', and
- `make-search-completion-function` helper to make completion functions.

Works, loads things (although takes some time with both Wikipedia and DuckDuckGo), shows fancy search suggestions!

(I'm going to add completion functions to `nx-search-engines` right after we merge this, as this is too exciting to ignore!)

How does it look for you?

Closes #79.

EDIT: Small readability fixes.
EDIT: "deafult" typo :)
EDIT: Add a linked issue.

jmercouris commented:
Absolutely blows my mind! Fantastic work Artyom!

That is just so damn cool!

Ambrevar commented:
Awesoooooooooooooooooooome!!! Thank you so much for working on this, will review as soon as possible!

P.S.: I see you had fun with theming :)

jmercouris commented:
It seems Pierre is not familiar with the venerable Laconia theme!

Ambrevar commented:
The related issue is #79, one of our oldest! ;)

jmercouris commented:
Resources: https://github.com/aartaka/nyxt-config/blob/master/style.lisp
https://github.com/aartaka/laconia-theme

Ambrevar commented:
I've tested it, it's absolutely amazing.
I suggest we merge this now and include it in 2-pre-release-7.
@jmercouris ?

Ambrevar commented on source/search-engine.lisp:
```diff
@@ -32,6 +36,28 @@ the empty string."))
+                          (function (string) list-of-strings))
+                make-search-completion-function))
+(defun make-search-completion-function (&key base-url
+                                          (request-function #'dex:get)
```
dex:get must use some proxy settings, no?
But which ones?  Does it make sense for the prompt buffer to use proxy settings of the current-buffer?
Should it have its own proxy settings?

Here is my suggestion: make a wrapper around `dex:get` that's uses the prompt buffer proxy.
Then the user can enable the proxy by adding proxy-mode to the prompt-buffer or not.
Does that make sense?

jmercouris commented:
I think we best leave it for actual 2.0.0. It is too much of a last minute thing in my opinion. Rather, let Artyom fix it in peace :-). What do you think?

Ambrevar commented:
Agreed!

aartaka commented:
Things left to do before merging:
- auto-completion using default search engine, e.g., when you type `salad` and it gives you `salad recipe`, `salad dressing` etc. even without the search engine shortcut.
- Proxying.
- Making auto-completion togglable: there are people as paranoidal as myself in regards to auto-completion, so we better make it disableable :D

Ambrevar commented:
Agreed on all points! :)

aartaka commented:
- Implemented:
  - Proxied auto-completion.
  - A buffer-located switch for auto-completion.
  - Auto-completion by default, for any query.
- Rebased over master to use the fix for the bug found in the process (11425b2300cbe295fc25beef7c52ebaf6a67cc20).
- Tested proxied downloads and disabled auto-completion. Works fine.

Mergeable, I believe :)

aartaka commented:
+ CI passes. Phew...

Ambrevar commented on source/search-engine.lisp:
```diff
@@ -32,6 +36,38 @@ the empty string."))
+                                             (proxy-address (current-buffer))
+                                             (quri:render-uri
+                                              (proxy-address (current-buffer))))))))
+    (apply #'dex:get url args)))
```
What about adding a `(proxy-url-string BUFFER)` function, then at the call site use

```lisp
(dex:get :proxy (proxy-url-string PROMPT-BUFFER))
```

?

By the way, shouldn't we rename `proxy-address` to `proxy-url` for consistency?

aartaka replied:
Hmmm. I'm not sure about the `proxy-url-string` helper. Let's leave it the way it is and see whether we need it in the future :)

Renaming `proxy-adress` to `proxy-url` sounds intuitive! I'll do that.

Ambrevar commented on source/buffer.lisp:
```diff
@@ -992,17 +1006,30 @@ Finally, if nothing else, set the `engine' to the `default-search-engine'."))
-                                                    :engine engine))
-                    engines))))
+  (flet ((make-engine-completions (engine input)
+           (when (and (and (current-buffer) (search-auto-complete-p (current-buffer)))
```
Nested `and` can be merged.

Ambrevar commented:
So how does one configure the prompt-buffer to be proxied?

aartaka commented:
There's no way as of now: I use `current-buffer` (which prompt-buffer cannot be) proxy to decide on proxying auto-completion. So, prompt-buffer is not involved. 

Should prompt-buffer have a proxy? I'm not sure -- prompt-buffer is made for internal interface use, after all.

jmercouris commented:
I agree with Artyoms assessment

jmercouris commented:
I understood that the prompt buffer will use the proxy settings of the current buffer

Ambrevar commented:
> I understood that the prompt buffer will use the proxy settings of the current buffer

But that's not correct.  For instance it will fail to do the right thing
when run from the *Help* buffer.

aartaka commented:
Fair point! While I don't feel right about adding proxy to the prompt-buffer, I see the problem.

jmercouris commented:
I would say don't add it to the prompt buffer, add it to the browser, that makes more sense. It can then be used for downloads with the lisp engine etc.

jmercouris commented:
I don't see that as likely, but a fair point. Being more specific has benefits.

It also has costs. Will the user know to set their proxy for every single aspect of the system?

Ambrevar commented:
> It also has costs. Will the user know to set their proxy for every single aspect of the system?

We can set the default value to that of (proxy *browser*).

Done! :)

Ambrevar commented:
Sorry, I wasn't very clear: We can set the default value _of the various local proxy slots_ to that of `(proxy *browser*)`.

jmercouris commented:
I see, that seems a good solution.

aartaka commented:
Hmmmm... How about yet another approach:
- Don't define auto-completion functions in Nyxt core.
- Leave proxied completion functions to be implemented by extension writers (because this logic actually belongs to extensions).
- Add `:request-args` keyword argument to the `make-search-completion-function` and use `dex:get` by default for `:request-function`. Example of usage:
```lisp
(make-search-completion-function
 :base-url "https://duckduckgo.com/ac/?q=~a"
 :request-args '(:proxy "socks5://localhost:9050")
 :processing-function
 #'(lambda (results)
     (mapcar #'cdar
             (json:decode-json-from-string results))))
```

Given that `nx-search-engines` is already destined to be a testbed for completion, I suggest completions functions (including proxied ones) are written there.

Rationale: 
- `prompt-buffer` is merely a display for search suggestions, so it doesn't make sense to add proxies to it.
- `buffer` makes more sense, given that most search-related slots (`search-engines`, `search-auto-complete-p`) are there. Why not use `buffer`'s `proxy` then? Because it's not every buffer that has proxies. Maybe instantiate buffers and enable `proxy-mode`? That's unreliable -- too much variables. Adding search-dedicated proxy slot doesn't sound right either.
- Adding `*browser*`-based proxy and inheriting that configuration is cool, but that's another global variable that supercedes a carefully crafted mode. Doesn't sound right.
- The mechanism is intended to be extensible, so why impose any implementation? If user wants to have things proxied, they can always override the `completion-function`. `make-search-completion-function` is intended for that, especially so with `:request-args`.

jmercouris commented:
I concur that the `make-search-completion-function` is the appropriate way for the user to customize this. It does not feel right that the `prompt-buffer` contain proxy information. It seems to me that would be outside of the prompt buffers concerns.

Ambrevar commented:
I agree with your last proposal, @aartaka! :)

aartaka commented:
> I agree with your last proposal, @aartaka! :)

I'll push it in an hour then!

aartaka commented:
Here it is: `:request-args` and `dex:get` used by default. Two things to discuss:
- Should I remove the default search engines' (DDG, Wiki) completion?
- `proxy-url` in `buffer` conflicts with `proxy-url` in `renderer-gtk`. Maybe rename GTK one?

jmercouris commented:
Thank you Artyom! I would leave the default search engine completion for now. I will look into what to do about the naming conflict!

jmercouris commented:
Perhaps it would make sense to name it %proxy-url in renderer-gtk.lisp, since it is an implementation detail.

jmercouris commented:
Or is it...? I guess it isn't. Decisions decisions...

jmercouris commented:
No, definitely it should be set to %proxy-url in `renderer-gtk.lisp`.

jmercouris commented:
@aartaka if you approve, I shall merge!

jmercouris commented:
I decided to just call it `gtk-proxy-url` in the end.

Ambrevar commented:
Looks good to me, except that I think we should expand add an example on how to add a proxy, and refer to this in the manual.

aartaka commented:
> Looks good to me, except that I think we should expand add an example on how to add a proxy, and refer to this in the manual.

Fair point. Done in 65092494ed1356baced050a207e76803db99cb7a. Feels all good to merge otherwise :)
